bit_vector/bit_vector_abc.py

```diff
@@ -0,0 +1,299 @@
+from abc import ABCMeta, abstractmethod
+from collections import namedtuple
+import typing as tp
+import functools as ft
+import weakref
+import warnings
+
+
+TypeFamily = namedtuple('TypeFamily', ['Bit', 'BitVector', 'Unsigned', 'Signed'])
+
+#I want to be able differentiate an old style call
+#BitVector(val, None) from BitVector(val)
+_MISSING = object()
+class AbstractBitVectorMeta(ABCMeta):
+    _class_cache = weakref.WeakValueDictionary()
+    _class_info  = weakref.WeakKeyDictionary()
+
+    def __call__(cls, value, size=_MISSING):
+        if cls.is_sized and size is not _MISSING:
+            raise TypeError('Cannot use old style construction on sized types')
+        elif cls.is_sized:
+            return super().__call__(value)
+        elif size is _MISSING or size is None:
+            if size is None:
+                warnings.warn('DEPRECATION WARNING: Use of BitVectorT(value, size) is deprecated')
+            if isinstance(value, AbstractBitVector):
+                size = value.size
+            elif isinstance(value, AbstractBit):
+                size = 1
+            elif isinstance(value, tp.Sequence):
+                size = max(len(value), 1)
+            elif isinstance(value, int):
+                size = max(value.bit_length(), 1)
+            elif hasattr(value, '__int__'):
+                size = max(int(value).bit_length(), 1)
+            else:
+                raise TypeError('Cannot construct {} from {}'.format(cls, value))
+        else:
+            warnings.warn('DEPRECATION WARNING: Use of BitVectorT(value, size) is deprecated')
+
+        return AbstractBitVectorMeta.__call__(cls[size], value)
+
+
+
+    def __new__(mcs, name, bases, namespace, **kwargs):
+        size = None
+        for base in bases:
+            if getattr(base, 'is_sized', False):
+                if size is None:
+                    size = base.size
+                elif size != base.size:
+                    raise TypeError("Can't inherit from multiple different sizes")
+
+        t = super().__new__(mcs, name, bases, namespace, **kwargs)
+        if size is None:
+            AbstractBitVectorMeta._class_info[t] = t, size
+        else:
+            AbstractBitVectorMeta._class_info[t] = None, size
+
+        return t
+
+    def __getitem__(cls, idx : int) -> 'AbstractBitVectorMeta':
+        try:
+            return AbstractBitVector._class_cache[cls, idx]
+        except KeyError:
+            pass
+
+        if not isinstance(idx, int):
+            raise TypeError()
+        if idx < 0:
+            raise ValueError('Size of BitVectors must be positive')
```

Comment:
Did we decide to allow size 0?

Reply:
I think we decided against it but forbidding them broke some tests and create weird corner cases for user code.  I think its better to allow them.

Reply:
Okay, let's leave it for now but keep on eye on whether this becomes a problem.